### PR TITLE
reconnection

### PR DESCRIPTION
**Pull request by visox**
- add logic to reconnect to db after connection break
- add tests
- add config file for in memory orientdb instance in test

### CODE REVIEW

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java [hunk 1]
```diff
@@ -128,6 +134,19 @@ public void makeActive() {
         }
     }
 
+    public void makeActive() {
+        makeActiveDb();
+
+        try {
+            this.database.getMetadata().getSchema().reload();
+        } catch (Exception _) {
+            this.database = new ODatabaseDocumentTx(this.database.getURL(), this.database.isKeepStorageOpen());
```
Comment (velo) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: what about the pool? don't need to recreate it as well?

just create a new database when pool is active doesn't seem to respect the pool.

**Comment by mpollmeier** [comment]
just discussed with michal (we're working together) that we'll optimise the check in `makeActive` to only be triggered after an actual failure. i.e. there won't be any unnecessary calls to the database as it's the case at the moment

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java (hunk 1 above)
Comment (visox) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: You are right but there are 2 problems, 
1. the database may have not been created using a pool (poll may be null)
2. `pool.acquire();` gives me a database with a failing connection (if there was a problem in the first place), so i really need to recreate the it by hand.

One performance change will follow.

**Comment by velo** [comment]
@mpollmeier nice... keep in mind the pool thing I commented!

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java [hunk 2]
```diff
@@ -128,30 +141,67 @@ public void makeActive() {
...
+
     @Override
     public Vertex addVertex(Object... keyValues) {
-        makeActive();
+        return executeWithConnectionCheck(() -> {
+            makeActive();
 
-        ElementHelper.legalPropertyKeyValueArray(keyValues);
-        if (ElementHelper.getIdValue(keyValues).isPresent()) throw Vertex.Exceptions.userSuppliedIdsNotSupported();
+            ElementHelper.legalPropertyKeyValueArray(keyValues);
```
Comment (mpollmeier) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: formatting

**Comment by visox** [comment]
The history is a bit messy, maybe i can stash the changes.

ALSO: so i tried to make use of the pool that is already there, there was a problem with that since `pool.acquire();` created a connection that did not reconnect. So in the current version there is a function `PoolRecreate` i pass the function to the graph and get a new pool -> new connection. Now to call `PoolRecreate` means that the pool is mutated in the factory and i am not sure if that is so good, any suggestions ?

**Comment by mpollmeier** [comment]
@velo any ideas? as michal said the current solution works but it's a bit ugly because it involves mutation, and if anyone is holding onto the pool reference that might get outdated quickly. 

I just checked and there seems to be some handling for this directly in orient itself, which would mean we don't actually need to do anything - if only we'd know why it's not being used at the moment...
https://github.com/orientechnologies/orientdb/blob/master/client/src/main/java/com/orientechnologies/orient/client/remote/OStorageRemote.java#L1758

**Comment by visox** [comment]
So i was debugging to see if i go through `OStorageRemote`, where `handleException` is defined, and indeed in stacktrace i can find `...at com.orientechnologies.orient.client.remote.OStorageRemote.getPhysicalClusterNameById` but that method does not handle the exception, it's not even in the `try` block that is there, so i guess it's by design (maybe wrong design ?) and up to us to handle that.

**Comment by velo** [comment]
May be, we should create a wrapper around the pool and in case the ODBPool is returning broken connections we can kill and recreate it...

**Comment by visox** [comment]
@lvca @laa any ideas/suggestions on this problem ? 

in short: we create a `OPartitionedDatabasePool pool` which can create a `ODatabaseDocumentTx` that connects fine, then after connection breaks and later works again `pool.acquire()` still returns a database that is not able to connect. One way is to recreate the whole pool but that is maybe too consuming and adds more complexity and mess to `org.apache.tinkerpop.gremlin.orientdb.OrientGraphFactory`
Last suggestion is to hide the reconnection and pool in a wrap which may reduce complexity and other problems in the client code.

**Comment by laa** [comment]
@visox Hi I think we may provide way to kill such database instance close.
Could you clarify what is result of calling of method database.isClosed() if connection is broken ?

**Comment by laa** [comment]
I mean in your case of course ))

**Comment by visox** [comment]
@laa 
```
import com.orientechnologies.orient.client.remote.OServerAdmin;
import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
import com.orientechnologies.orient.server.OServerMain;

public class ODatabaseDocumentTx_isClosed_Test {
    private static String url = "remote:localhost";
    private static String user = "root";
    private static String pass = "somepass";
    private static String dbName = "db-test-failing";
    private static String dbType = "graph";
    private static String dbStorage = "memory";

    private static void createDb() throws Exception {
        new OServerAdmin(url)
                .connect(user, pass)
                .createDatabase(dbName, dbType, dbStorage).close();
    }

    private static void startServer() throws Exception { OServerMain.main(new String[0]); }
    private static void stopServer() { OServerMain.server().shutdown(); }

//config/orientdb-server-config.xml needs to be present for OServerMain

    public static void main(String [] args) throws Exception {
        startServer();
        createDb();

        ODatabaseDocumentTx db = new ODatabaseDocumentTx(url + "/" + dbName);
        db.open(user, pass);

        stopServer();

        try {
            boolean closed = db.isClosed();
            System.out.println(closed);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
```
`false` (and no exception)

**Comment by laa** [comment]
Cool , could you create issue that we have to remove databases with closed connections from the pool . I will provide fix for next 2.1 hotfix ?

**Pull request merged**